rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,164 @@
+
+## Scope
+
+### In scope
+
+This RFC is purely about developing a bytecode VM for VRL.
+
+### Out of scope
+
+Any other performance issues relating to VRL will be discussed in
+[rfcs/2021-10-14-9811-vrl-performance.md](https://github.com/vectordotdev/vector/pull/9812)
+
+## Proposal
+
+### Implementation
+
+Vrl compiles to an AST that is then walked during resolution. Each node in that
+tree is boxed and stored in disparate regions of memory. As a result walking
+the tree means that the CPU caches must be constantly swapped.
+
+Instead we can create a bytecode VM to store the execution of the Vrl program.
+
+Bytecode is essentially a big enum of instructions (that are represented by an integer
+using the [num-derive](https://crates.io/crates/num-derive),
+[num-traits](https://crates.io/crates/num-traits) crates):
+
+```rust
+#[derive(FromPrimitive, ToPrimitive, Copy, Clone, Debug, PartialEq, Eq)]
+pub enum OpCode {
+    Return = 255,
+    Constant,
+    Negate,
+    Add,
+    Subtract,
+    Multiply,
+    Divide,
+    Print,
+    Not,
+    Greater,
+    GreaterEqual,
+    Less,
+    LessEqual,
+    NotEqual,
+    Equal,
+    Pop,
+    JumpIfFalse,
+    Jump,
+    SetPath,
+    GetPath,
+    Call,
+    ...
+}
+```
+
+The Vm is a struct comprising of the following fields:
+
+```rust
+#[derive(Clone, Debug, Default)]
+pub struct Vm {
+    instructions: Vec<usize>,
+    values: Vec<Literal>,
+    targets: Vec<Variable>,
+    stack: Vec<Value>,
+    ip: usize,
+}
+```
+
+- instructions
+
+The instructions field is a `Vec` of `OpCode` cast to a usize. The reason for
+the cast is because not all instructions are `OpCode`. For example the
+instructions `[.., Constant, 12, ..]` when evaluated will load the constant
+stored in the `values` `Vec` that is found in position 12 onto the stack.
+
+- values
+
+A list of constant values found in the program. Since the bytecode only
+contains integers any actual values must be stored here. This also allows
+literals to be deduped.
+
+- targets
+
+A list of paths used in the program, similar to `values`.
+
+- stack
+
+The Vm is a stack based Vm. Every expression that is evaluated pushes the
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- ip
+
+The instruction pointer points to the next instruction to evaluate.
+
+With each node of the AST compiled down to just a few bytes and all
+instructions held in contiguous memory evaluation of the program should be able
+to take full advantage of the CPU cache which should result in much faster
+execution.
+
+
+#### Calling functions
+
+Calling functions in the stdlib will be a case of evaluating each parameter
+with the results pushed onto the stack.
+
+Since Vrl allows for named parameters and optional parameters, the compiler
+will need to ensure the bytecode evaluates each parameter in a specific order
+(likely the order declared in the function). Bytecode will need to be emmitted
```

Comment:
This needs some clarification. Are you talking about the order declared in the _function implementation_, or the _function call_?

I haven't dug into this too deep yet, but it's important to note that the current implementation allows positional and named parameters to be mixed and matched, and ordering of named parameters is irrelevant.

That is, these are the same in VRL:

```coffee
encode_base64(value: .value, padding: true, charset: "...")
encode_base64(.value, true, "...")
encode_base64(charset: "...", padding: true, value: .value)
encode_base64(padding: true, .value, charset: "...")
# etc...
```

Reply:
This would have to be as declared in the function implementation. The alternative would be to push named parameters onto the stack along with their name and let the callee put things in the right order. The alternative described here is _much_ more efficient.

Reply:
I like the idea of keeping the names with the values. 

I'm thinking it might be an idea to have a separate stack for parameters, this stack would be a `Vec<(&'static str, Value)>`. We have an OpCode for moving a value from the main stack to the parameter stack tagged with the name.

The function call will pull the values from this stack and pass the parameters to the function.

Reply:
In the Python interpreter, at least, this method is a major performance cost since the called function then needs to decode the parameter ordering. Since we know the required parameters (because the stdlib is built in), we can fix the ordering and push/pop exactly what is needed to make it fastest.

Reply:
On second thoughts, I think adding the parameter name to the OpCode increases `size_of::<OpCode>()` to 24. We could include a similar lookup into a table, but I think given that we have all the information we need based on the position the parameters are declared in the function, we can work perfectly well just by position in the stack (with a placeholder for unspecified optional parameters).